File: Oriented-Object Programming/Contador/src/projeto_bimestral/Quicksort.java

package projeto_bimestral;

public class Quicksort {
    

    int partition(int arr[], int left, int right) {

        int i = left, j = right;

        int tmp;

        int pivot = arr[(left + right) / 2];

        while (i <= j) {

            while (arr[i] < pivot) {
                i++;
            }

            while (arr[j] > pivot) {
                j--;
            }

            if (i <= j) {

                tmp = arr[i];

                arr[i] = arr[j];

                arr[j] = tmp;

                i++;

                j--;

            }

        };

        return i;

    }

    void quicksort(int arr[], int left, int right) {

        int index = partition(arr, left, right);

        if (left < index - 1) {
            quicksort(arr, left, index - 1);
        }

        if (index < right) {
            quicksort(arr, index, right);
        }

    }

    
    public static void quick_sorted(int[] x) {
        if (x.length > 1) {
            new Quicksort().quicksort(x, 0, x.length - 1);
          
        }
    }

}

File: rust/algorithms-library/src/sorting/quicksort.rs
pub fn quicksort<T>(arr: &mut [T]) -> ()
where
    T: PartialOrd<T>,
{
    if arr.len() <= 1 {
        return;
    }

    let pivot_index: usize = partition(arr);

    let (left, right) = arr.split_at_mut(pivot_index);

    quicksort(left);
    quicksort(&mut right[1..]);
}

fn partition<T>(arr: &mut [T]) -> usize
where
    T: PartialOrd<T>,
{
    let pivot_index: usize = arr.len() / 2;

    arr.swap(pivot_index, arr.len() - 1);

    let mut index = 0;

    for i in 0..arr.len() - 1 {
        if arr[i] < arr[arr.len() - 1] {
            arr.swap(i, index);

            index += 1;
        }
    }

    arr.swap(index, arr.len() - 1);
    index
}
